coalib/bears/Bear.py

```diff
@@ -305,6 +314,36 @@ def log_message(self, log_message, timestamp=None, **kwargs):
     def run(self, *args, dependency_results=None, **kwargs):
         raise NotImplementedError
 
+    def setup_profile(self, profiler=None):
+        filename = '{}_{}.prof'.format(self.section.name, self.name)
+        if self.profile is True:
+            profiler.dump_stats(filename)
+        elif os.path.exists(self.profile) and os.path.isdir(self.profile):
+            profiler.dump_stats(os.path.join(self.profile, filename))
+        elif not os.path.isfile(self.profile):
+            os.mkdir(os.path.join(self.profile))
+            profiler.dump_stats(os.path.join(self.profile, filename))
+        else:
+            print('directory and filename cannot be same', file=sys.stdout)
```

Comment:
@Makman2  
1. first it will check whether it will check the value of `self.profile` if its `True` then it will dump in the current working directory,
2. else  it will check whether the path given or `directory name` or in `self.profile`  already exists and also check it is a directory if yes then it will dump the profiler files in that directory.

3. else it will check whether the path given or `directory name` is already a `file` (not directory) if it is not a file then it will create a `directory` with the name passed in `self.profile` or the `path passed` and dump the profiler files in this directory.
3. else last case which left is user is passing the `filename` or `filepath` in `self.profile` then it becomes obvious that it is not possible to dump profiler files in `file` (I will raise the exception in this case instead of printing)

Reply:
IMO this also get the ouput we want and also cover the case where user pass the path or filename which is actually a file not a directory it will give a proper error which user can understand easily
i.e., `FileExistsError: [Errno 17] File exists: '/home/vaibhav/test/my.txt'`
But if `/home/vaibhav/test/my.txt` is a directory then it will not give any error 
, what I was not wanted was to halt the coala normal execution and I am hoping this is the thing you wanted?
```
    def setup_profile(self, profiler=None):
        filename = '{}_{}.prof'.format(self.section.name, self.name)
        if self.profile is True:
            profiler.dump_stats(filename)
        else:
            if os.path.exists(self.profile) and os.path.isdir(self.profile):
                profiler.dump_stats(os.path.join(self.profile, filename))
            else:
                os.mkdir(os.path.join(self.profile))
                profiler.dump_stats(os.path.join(self.profile, filename))
```

Reply:
It's okay to halt coala's execution, if the user requested to profile code, but you can't profile because of such filename issues, you want to rather stop and make the user fix it. Otherwise you do a full coala run (that depending on the setup could take minutes) without profiling which wastes time.

Reply:
And yeah, your code looks better :+1: However one doubt: So if the user specifies a path in `self.profile`, we want to make it a directory, right?
Two improvements:
1. In the first else branch, you execute `profiler.dump_stats(os.path.join(...))` regardless on which branch, that can be moved outside the if ;)
2. since in the end you will always do `dump_stats`, and the only thing that changes is the filename, how about doing:

```python
    def setup_profile(self, profiler=None):
        filename = '{}_{}.prof'.format(self.section.name, self.name)
        if self.profile is True:
            path = filename
        else:
            if not (os.path.exists(self.profile) and os.path.isdir(self.profile)):
                os.mkdir(os.path.join(self.profile))
            
            path = os.path.join(self.profile, filename)            

    profiler.dump_stats(path)
```